[DOCS-10436] adding HTTP tests to Mobile app testing (#28334)
DataDog/documentation: content/en/synthetics/mobile_app_testing/mobile_app_tests/steps.md

{{% /tab %}}
{{< /tabs >}}

### Interactions

In addition to automatically recording steps based on your device assertions, you can also manually create steps by clicking **Interaction**. 

> {{< img src="mobile_app_testing/test_steps/mobile_app_interaction.png" alt="Choose an action type to add an interaction step" style="width:60%;" >}}

Author: Ah yes I'll fix that screen shot

By default, mobile app tests wait for a page to be fully loaded before performing a step or the next step with a timeout of 60 seconds. This additional time is systematically added to **every run** of your mobile app test's recording.

#### Toggle Wi-Fi

This action allows you to enable or disable Wi-Fi within your test to monitor how your application performs with or without internet access.

{{< img src="mobile_app_testing/test_steps/toggle_wifi.png" alt="Screenshot of the Toggle Wi-Fi interaction step" style="width:60%" >}}

#### HTTP requests

You can run HTTP requests as part of your mobile app tests. If you are using [restrictive networks][8], ensure the required IP address ranges are added to your to allowlist to enable HTTP steps to run in your Synthetic Monitoring Mobile tests.

**Note**: You can copy an HTTP step from a Synthetic Monitoring Mobile test and re-use it into a Synthetic Monitoring Browser test. However, copying steps from a Browser test to a Mobile test is not supported.

{{< img src="mobile_app_testing/test_steps/http_interaction.png" alt="Mobile app testing interactions Run HTTP test" style="width:70%;" >}}

To define your HTTP request:

> 1. Select **Interaction**, then **Run HTTP Test**. Enter the URL you wish to test.

Reviewer: ```suggestion
1. Enter the URL you wish to test.
```
Same note as above with these first clicks are kind of already covered 🙂

Author: fixed!

2. Optionally, specify **Advanced Options**:
   
   {{< tabs >}}

   {{% tab "Request Options" %}}

>    * **Follow redirects**: Check to have your HTTP test follow up to ten redirects when performing the request.
>    * **Ignore server certificate error**: Check to have your HTTP test go on with connection even if there are errors when validating the SSL certificate.

Reviewer: ```suggestion
   * **Follow redirects**: Select this option to have your HTTP test follow up to ten redirects when performing the request.
   * **Ignore server certificate error**: Select this option to have your HTTP test go on with connection even if there are errors when validating the SSL certificate.
```

   * **Cookies**: Define cookies to add to your HTTP request. Set multiple cookies using the format `<COOKIE_NAME1>=<COOKIE_VALUE1>; <COOKIE_NAME2>=<COOKIE_VALUE2>`.

   {{% /tab %}}

   {{% tab "Authentication" %}}

   * **Client certificate**: Authenticate through mTLS by uploading your client certificate and the associated private key.
   * **HTTP Basic Auth**: Add HTTP basic authentication credentials.
   * **Digest Auth**: Add Digest authentication credentials. 
   * **AWS Signature**: Add AWS Access Key ID and Secret Access Key.
   * **NTLM**: Add NTLM authentication credentials. Support both NTLMv2 and NTLMv1.
   * **OAuth 2.0**: Select a Grant Type (Client credentials, or Resource owner password).

   {{% /tab %}}

   {{% tab "Query Parameters" %}}

   * **Encode parameters**: Add the name and value of query parameters that require encoding. 

   {{% /tab %}}

   {{% tab "Request Body" %}}

   * **Body type**: Select the type of the request body (`text/plain`, `application/json`, `text/xml`, `text/html`, `application/x-www-form-urlencoded`, `GraphQL`, or `None`) you want to add to your HTTP request.

>    * **Request body**: Add the content of your HTTP request body. The request body is limited to a maximum size of 50 kilobytes.

Reviewer: ```suggestion
   * **Request body**: Add the content of your HTTP request body. The request body is limited to a maximum size of 50 KB.
```

[1]: /data_security/synthetics
   {{% /tab %}}

   {{< /tabs >}}
   </br>
3. Click **Send** to try out the request configuration. A response preview appears.

{{< img src="mobile_app_testing/test_steps/http_mobile_request.png" alt="Make HTTP Request" style="width:80%;" >}}

##### Add assertions

Assertions define what an expected test result is. After you click **Send**, basic assertions on `status code`, `response time`, and `header` `content-type` are added based on the test response. Assertions are optional for HTTP steps in browser tests.

| Type            | Operator                                                                                                               | Value type                                               |
|-----------------|------------------------------------------------------------------------------------------------------------------------|----------------------------------------------------------|

> | `body`          | `contains`, `does not contain`, `is`, `is not`, <br> `matches`, `does not match`, <br> [`jsonpath`][11], [`xpath`][12] | _String_ <br> _[Regex][13]_ <br> _String_, _[Regex][13]_ |

Reviewer: Thanks for adding the URL for [13]! Looks like the ones for [11] and [12] are still missing, but once they're in there, we should be ready to boogie 🙂

Author: oh thank you!! Not sure how I missed this :)

Reviewer: Easy to miss for sure!

> | header        | `contains`, `does not contain`, `is`, `is not`, <br> `matches`, `does not match`                       | _String_ <br> _[Regex][13]_                                      |

Reviewer: These links are broken - it looks like all you need to do is add in the link references at the bottom of the page and it'll fix it 🙂

Reviewer: Same notes as above about formatting the items in the Type column and the breaks! Just figured that if you want to switch up the breaks in the other table that you might just want to duplicate it here.

Author: let me check live, if they are broken in live, that is an oversight. thanks for catching.

Author: fixed!

| response time | `is less than`                                                                                         | _Integer (ms)_                                                  |
| status code   | `is`, `is not`                                                                                         | _Integer_                                                      |

HTTP requests can decompress bodies with the following `content-encoding` headers: `br`, `deflate`, `gzip`, and `identity`.

- If a test does not contain an assertion on the response body, the body payload drops and returns an associated response time for the request within the timeout limit set by the Synthetics Worker.

- If a test contains an assertion on the response body and the timeout limit is reached, an `Assertions on the body/response cannot be run beyond this limit` error appears.

{{< img src="synthetics/browser_tests/assertions.png" alt="Define assertions for your browser test to succeed or fail on" style="width:80%;" >}}

You can create up to 20 assertions per step by clicking **New Assertion** or by clicking directly on the response preview.

##### Extract a variable from the response

> Optionally, extract a variable from the response of your HTTP request by parsing its response headers or body. The value of the variable updates each time the HTTP request step runs. Once created, this variable can be used in the [following steps](#use-variables) of your browser test.

Reviewer: Same suggested rephrasing as above 🙂

To start parsing a variable, click **Extract a variable from response content**:

1. Enter a **Variable Name**. Your variable name can only use uppercase letters, numbers, and underscores and must have at least three characters.
2. Decide whether to extract your variable from the response headers or the response body.

>    * Extract the value from **response header**: use the full response header of your HTTP request as the variable value or parse it with a [`regex`][13].
>    * Extract the value from **response body**: use the full response body of your HTTP request as the variable value or parse it with a [`regex`][13], a [`JSONPath`][11], or a [`XPath`][12].

Reviewer: These links are also broken - again, I think they'll be fine once the corresponding links are there. Also, I popped in the same suggestion to remove the articles again.
```suggestion
   * Extract the value from **response header**: use the full response header of your HTTP request as the variable value or parse it with [`regex`][13].
   * Extract the value from **response body**: use the full response body of your HTTP request as the variable value or parse it with [`regex`][13], [`JSONPath`][11], or [`XPath`][12].
```

   * **Request headers**: Define headers to add to your HTTP request. You can also override the default headers (for example, the `user-agent` header).
   * **Cookies**: Define cookies to add to your HTTP request. Set multiple cookies using the format `<COOKIE_NAME1>=<COOKIE_VALUE1>; <COOKIE_NAME2>=<COOKIE_VALUE2>`.

   {{% /tab %}}

   {{% tab "Authentication" %}}

   * **Client certificate**: Authenticate through mTLS by uploading your client certificate and the associated private key.
   * **HTTP Basic Auth**: Add HTTP basic authentication credentials.
   * **Digest Auth**: Add Digest authentication credentials. 
   * **AWS Signature**: Add AWS Access Key ID and Secret Access Key.
   * **NTLM**: Add NTLM authentication credentials. Support both NTLMv2 and NTLMv1.
   * **OAuth 2.0**: Select a Grant Type (Client credentials, or Resource owner password)

   {{% /tab %}}

   {{% tab "Query Parameters" %}}

   * **Encode parameters**: Add the name and value of query parameters that require encoding. 

   {{% /tab %}}

   {{% tab "Request Body" %}}

>    * **Body type**: Select the type of the request body (`text/plain`, `application/json`, `text/xml`, `text/html`, `application/x-www-form-urlencoded`, `GraphQL`, or `None`) you want to add to your HTTP request.

Reviewer 2: We don't support file upload on Mobile HTTP steps, so we should leave the body type as is. (the other two added to browser in my comment involve body file upload). All good here

   * **Request body**: Add the content of your HTTP request body. The request body is limited to a maximum size of 50 kilobytes.

   {{% /tab %}}

   {{% tab "Proxy" %}}

   * **Proxy URL**: Specify the URL of the proxy the HTTP request should go through (`http://<YOUR_USER>:<YOUR_PWD>@<YOUR_IP>:<YOUR_PORT>`).
   * **Proxy Header**: Add headers to include in the HTTP request to the proxy.

   {{% /tab %}}
  
   {{% tab "Privacy" %}}

   * **Do not save response body**: Select this option to prevent the response body from being saved at runtime. This helps ensure no sensitive data is displayed in your test results, but it can make failure troubleshooting more difficult. For full security recommendations, see [Synthetic Monitoring Data Security][1].

[1]: /data_security/synthetics
   {{% /tab %}}

   {{< /tabs >}}
   </br>
3. Click **Send** to try out the request configuration. A response preview appears.

{{< img src="mobile_app_testing/test_steps/http_mobile_request.png" alt="Make HTTP Request" style="width:80%;" >}}

> 

Reviewer 2: We should add the `Add assertions` and `Extract a variable` subtitles for Mobile HTTP steps here, like suggested in the other comment for Browser.

Author: added
